File: output_20260318_091315/claude-opus/c/c/Struct/pred.c
#include "stdlib.h"

/*@ predicate account(struct account *a; int limit, int balance) =
      a->limit |-> limit &*& a->balance |-> balance;
 @*/

struct account
{
    int limit;
    int balance;
};

struct account *create_account(int limit)
    //@ requires true;
    //@ ensures account(result, limit, 0);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->limit = limit;
    myAccount->balance = 0;
    
    return myAccount;
}

int account_get_balance(struct account *myAccount)
    //@ requires account(myAccount, ?limit, ?balance);
    //@ ensures account(myAccount, limit, balance) &*& result == balance;
{
    
    int result = myAccount->balance;
    
    return result;
}

void account_deposit(struct account *myAccount, int amount)
    //@ requires account(myAccount, ?limit, ?balance);
    //@ ensures account(myAccount, limit, balance + amount);
{
    
    myAccount->balance += amount;
    
}

int account_withdraw(struct account *myAccount, int amount)
    //@ requires account(myAccount, ?limit, ?balance);
    /*@ ensures
          account(myAccount, limit, balance - result) &*&
          result == (balance - amount < limit ? balance - limit : amount);
    @*/
{
    
    int result = myAccount->balance - amount < myAccount->limit ? myAccount->balance - myAccount->limit : amount;
    myAccount->balance -= result;
    
    return result;
}

void account_dispose(struct account *myAccount)
    //@ requires account(myAccount, _, _);
    //@ ensures true;
{
    
    free(myAccount);
}

int main()
    //@ requires true;
    //@ ensures true;
{
    struct account *myAccount = create_account(-100);
    account_deposit(myAccount, 200);
    int w1 = account_withdraw(myAccount, 50);
    assert(w1 == 50);
    int b1 = account_get_balance(myAccount);
    assert(b1 == 150);
    int w2 = account_withdraw(myAccount, 300);
    assert(w2 == 250);
    int b2 = account_get_balance(myAccount);
    assert(b2 == -100);
    account_dispose(myAccount);
    return 0;
}

File: output_repair/claude-opus/sample_3/repair_2/rust/rust/purely_unsafe/solutions/pred.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Account {
    limit: i32,
    balance: i32,
}

predicate account_owned(p: *mut Account, limit: i32, balance: i32) =
    Owned(p) && (*p).limit == limit && (*p).balance == balance;

impl Account {
    unsafe fn create(limit: i32) -> *mut Account
        ensures account_owned(result, limit, 0)
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).limit = limit;
        (*my_account).balance = 0;

        my_account
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32
        requires exists<l: i32, b: i32> account_owned(my_account, l, b)
        ensures exists<l: i32, b: i32> account_owned(my_account, l, b) && result == b
    {
        let result = (*my_account).balance;

        result
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32)
        requires exists<l: i32, b: i32> account_owned(my_account, l, b)
        ensures exists<l: i32, b: i32> account_owned(my_account, l, b + amount)
    {
        (*my_account).balance += amount;
    }

    unsafe fn withdraw(my_account: *mut Account, amount: i32) -> i32
        requires exists<l: i32, b: i32> account_owned(my_account, l, b)
        ensures exists<l: i32, b: i32> account_owned(my_account, l, b - result)
    {
        let result = if (*my_account).balance - amount < (*my_account).limit {
            (*my_account).balance - (*my_account).limit
        } else {
            amount
        };
        (*my_account).balance -= result;

        result
    }

    unsafe fn dispose(my_account: *mut Account)
        requires exists<l: i32, b: i32> account_owned(my_account, l, b)
        ensures emp
    {
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }
}

fn main() {
    unsafe {
        let my_account = Account::create(-100);
        Account::deposit(my_account, 200);
        let w1 = Account::withdraw(my_account, 50);
        std::hint::assert_unchecked(w1 == 50);
        let b1 = Account::get_balance(my_account);
        std::hint::assert_unchecked(b1 == 150);
        let w2 = Account::withdraw(my_account, 300);
        std::hint::assert_unchecked(w2 == 250);
        let b2 = Account::get_balance(my_account);
        std::hint::assert_unchecked(b2 == -100);
        Account::dispose(my_account);
    }
}